File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        cyclesort(nums);
        
        for(int i = 0; i <nums.length;i++){
            if(nums[i]!=i+1){
               return nums[i];
            }
        }
        return-1;
    }

    public void cyclesort (int arr[]){
        int n = arr.length;
        int index= 0;
        while(index<n){
            int element = arr[index];
            int actualpos = element-1;
            if(arr[index]<n && arr[index]!= arr[actualpos]){
                swap(arr, index, actualpos);
            } else {
                index++;
            }
        }
    }

    public void swap (int arr[], int i , int j){
        int temp = arr[i];
        arr[i]= arr[j];
        arr[j]= temp;
    }
}

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function(nums) {
    for (let index = 0; index < nums.length; index++) {
        const element =Math.abs(nums[index]);
        const actualpos = element-1;
        if (nums[actualpos] < 0) {
            return element;
        }
        nums[actualpos] = -1*nums[actualpos];
    }
    return -1;
};